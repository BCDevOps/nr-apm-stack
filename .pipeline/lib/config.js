'use strict';
process.on('unhandledRejection', (reason) => {
  console.log(reason);
  process.exit(1);
});

module.exports = class {
  constructor(options) {
    this.options = options
  }
  build() {
    const options = this.options
    const name = 'nress'
    //const phases = ['build', 'dev', 'test', 'prod']
    const changeId = options.pr //aka pull-request

    const phases0 = {
      transient: {build: true                          , dev: true                                                            , test: false                                                         , prod: false},
      name:      {build: `${name}`                     , dev: `${name}`                                                       , test: `${name}`                                                     , prod: `${name}`},
      suffix:    {build: `-build-${changeId}`          , dev: `-dev-${changeId}`                                              , test: `-test`                                                       , prod: `-prod`},
      phase:     {build: 'build'                       , dev: 'dev'                                                           , test: 'test'                                                        , prod: 'prod'},
      keycloak:  {
        build: null,
        dev: {baseURL: process.env.KEYCLOAK_BASEURL, realmName:process.env.KEYCLOAK_REALM_NAME, clientId: process.env.KEYCLOAK_CLIENT_ID, clientSecret: process.env.KEYCLOAK_CLIENT_SECRET},
        test: {baseURL: process.env.KEYCLOAK_BASEURL, realmName:process.env.KEYCLOAK_REALM_NAME, clientId: process.env.KEYCLOAK_CLIENT_ID, clientSecret: process.env.KEYCLOAK_CLIENT_SECRET},
        prod: {baseURL: process.env.KEYCLOAK_BASEURL, realmName:process.env.KEYCLOAK_REALM_NAME, clientId: process.env.KEYCLOAK_CLIENT_ID, clientSecret: process.env.KEYCLOAK_CLIENT_SECRET},
      },
    };
    const phases = {};
    // Pivot configuration table, so that `phase name` becomes a top-level property
    // { namespace: { build: '-tools',  dev: '-dev'}}   ->  { build: { namespace: '-tools' }, dev: { namespace: '-dev' } }
    Object.keys(phases0).forEach((properyName) => {
      const property = phases0[properyName];
      Object.keys(property).forEach((phaseName) => {
        phases[phaseName] = phases[phaseName] || {};
        phases[phaseName][properyName] = property[phaseName];
      });
    });

    return {phases, options, environments: ['dev', 'test', 'prod']}
  }
}