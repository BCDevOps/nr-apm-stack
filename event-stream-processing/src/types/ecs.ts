import { Event } from './event';

export interface esc {
  Event: Event;
}
