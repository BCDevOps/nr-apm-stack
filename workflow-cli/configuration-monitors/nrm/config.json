{
  "type": "onpremise-monitor",
  "prefix": "nrids_agent_fluentbit_"
}