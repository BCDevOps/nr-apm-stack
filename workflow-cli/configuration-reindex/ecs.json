{
  "source": {
    "indexPrefix": "nrm-logs-access-"
  },
  "dest": {
    "indexPrefix": "nrm-access-external-"
  },
  "prototype": {
    "conflicts": "proceed",
    "source": {},
    "dest":{
      "op_type":"create"
    },
    "script": {
      "lang": "painless",
      "inline": ""
    }
  }
}
