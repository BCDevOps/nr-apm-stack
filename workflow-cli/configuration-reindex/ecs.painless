if (ctx._source.event.dataset == "apache.access") {
  if (ctx._source.service == null) {
    ctx._source.service = new HashMap();
  }
  if (ctx._source.offset != null) {
    ctx._source.event.sequence = ctx._source.remove("offset");
  }
  if (ctx._source.apache != null) {
    ctx._source.remove("apache");
  }
  if (ctx._source.message != null) {
    ctx._source.event.original = ctx._source.remove("message");
  }
  ctx._source.log.file.remove("extension");
  ctx._source.log.file.remove("name");
  ctx._source.log.file.remove("directory");
  ctx._source.remove("@event");
  if (ctx._source.url != null && ctx._source.url.file != null) {
    ctx._source.url.remove("file");
  }
  if (ctx._source.url != null && ctx._source.url.uri != null) {
    ctx._source.url.remove("uri");
  }
  if (ctx._source.url != null && ctx._source.url.query_v != null) {
    ctx._source.url.remove("query_v");
  }
  if (ctx._source.url != null && ctx._source.url.query_kv != null) {
    ctx._source.url.remove("query_kv");
  }
  if (ctx._source.url != null && ctx._source.url.query_k != null) {
    ctx._source.url.remove("query_k");
  }
  if (ctx._source.url != null && ctx._source.url.path_param != null) {
    ctx._source.url.remove("path_param");
  }
  if (ctx._source.labels.application != null) {
    ctx._source.labels.project = ctx._source.labels.remove("application");
  }
  if (ctx._source.labels.env_alias != null) {
    ctx._source.labels.environment_alias = ctx._source.labels.remove("env_alias");
  }
  if (ctx._source.labels.format != null) {
    ctx._source.labels.remove("format");
  }
  if (ctx._source.labels.context != null) {
    ctx._source.url.context = ctx._source.labels.remove("context");
  }
  if (ctx._source.labels.env != null) {
    ctx._source.service.environment = ctx._source.labels.remove("env");
    if (ctx._source.service.environment == "staging1" || ctx._source.service.environment == "staging2" || ctx._source.service.environment == "staging3") {
      ctx._source.service.environment = "staging";
    }
  }
  if (ctx._source.client != null) {
    ctx._source.source = ctx._source.remove("client");
  }
  if (ctx._source.http != null && ctx._source.http.request != null && ctx._source.http.request.referrer != null) {
    ctx._source.http.request.referrer = ctx._source.http.request.referrer.original;
  }
  if (ctx._source.http != null && ctx._source.http.request != null) {
    ctx._source.http.request.remove("line");
  }
  ctx._source.event.remove("original_timestamp");
  if (ctx._source.tls != null && ctx._source.tls.cypher != null) {
    ctx._source.tls.cipher = ctx._source.tls.remove("cypher");
  }

  if (ctx._source.user_agent != null && ctx._source.user_agent.engine != null) {
    ctx._source.user_agent.remove("engine");
  }

  if (ctx._source.user_agent != null && ctx._source.user_agent.cpu != null) {
    ctx._source.user_agent.remove("cpu");
  }

  if (ctx._source.user_agent != null && ctx._source.user_agent.device != null && ctx._source.user_agent.device.name == null) {
    ctx._source.user_agent.remove("device");
  }

  if (ctx._source.user_agent != null && ctx._source.user_agent.browser != null) {
    if (ctx._source.user_agent.browser.name != null && ctx._source.user_agent.name == null) {
      ctx._source.user_agent.name = ctx._source.user_agent.browser.name;
    }
    if (ctx._source.user_agent.browser.version != null && ctx._source.user_agent.version == null) {
      ctx._source.user_agent.version = ctx._source.user_agent.browser.version;
    }
    ctx._source.user_agent.remove("browser");
  }

  if (ctx._source.timestamp != null) {
    ctx._source.remove("timestamp");
  }

  if (ctx._source.event != null && ctx._source.event.kinesis != null) {
    ctx._source.kinesis = ctx._source.event.remove("kinesis");
  }
}
if (ctx._source.event.dataset == "express.access") {
  if (ctx._source.host.namespace != null) {
    ctx._source.host.name = ctx._source.host.remove("namespace");
  }

  if (ctx._source.http.request.referrer.uri != null) {
    ctx._source.http.request.referrer = ctx._source.http.request.referrer.uri;
  }

  if (ctx._source.http.response.time != null) {
    ctx._source.event.duration = ctx._source.http.response.remove("time") * 1000000;
  }
}

if (ctx._source.event.dataset == "node.css.http") {
  if (ctx._source.service == null) {
    ctx._source.service = new HashMap();
  }
  if (ctx._source.host == null) {
    ctx._source.host = new HashMap();
  }

  if (ctx._source.message != null) {
    ctx._source.event.original = ctx._source.remove("message");
  }
  if (ctx._source.timestamp != null) {
    ctx._source.remove("timestamp");
  }
  if (ctx._source.labels.env != null) {
    ctx._source.service.environment = ctx._source.labels.remove("env");
    if (ctx._source.service.environment == "staging1" || ctx._source.service.environment == "staging2" || ctx._source.service.environment == "staging3") {
      ctx._source.service.environment = "staging";
    }
  }
  if (ctx._source.labels.application != null) {
    ctx._source.labels.project = ctx._source.labels.remove("application");
    ctx._source.service.name = ctx._source.labels.project;
  }
  if (ctx._source.kubernetes != null) {
    if (ctx._source.kubernetes.namespace_name != null) {
      ctx._source.host.name = ctx._source.kubernetes.remove("namespace_name");
    }
    if (ctx._source.kubernetes.pod_ip != null) {
      ctx._source.host.ip = ctx._source.kubernetes.remove("pod_ip");
    }
    if (ctx._source.kubernetes.pod_name != null) {
      ctx._source.host.hostname = ctx._source.kubernetes.remove("pod_name");
    }
    ctx._source.remove("kubernetes");
  }

  if (ctx._source.log.file.offset != null) {
    ctx._source.event.sequence = ctx._source.log.file.remove("offset");
  }
  ctx._source.log.file.remove("extension");
  ctx._source.log.file.remove("directory");
  ctx._source.log.file.remove("name");
  ctx._source.remove("ip");
  ctx._source.event.type = "access";
  if (ctx._source.client != null) {
    ctx._source.source = ctx._source.remove("client");
  }
  if (ctx._source.http != null && ctx._source.http.response.time != null) {
    ctx._source.event.duration = ctx._source.http.response.remove("time") * 1000000;
  }
  if (ctx._source.http == null) {
    ctx['op'] = 'delete';
  }
}
