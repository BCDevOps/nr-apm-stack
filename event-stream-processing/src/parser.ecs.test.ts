import { create as buildContainer, TYPES } from "./inversify.config";
import { Randomizer } from "./randomizer.isvc";
import { ParserEcs } from "./parser.ecs.svc";

const myContainer = buildContainer()

beforeEach(() => {
    myContainer.snapshot();
    myContainer.rebind<Randomizer>(TYPES.Randomizer).toConstantValue({randomBytes:(size: number)=>{ return Buffer.from([0x62, 0x75, 0x66, 0x66, 0x65, 0x72])}})
})

afterEach(() => {
    myContainer.restore();
})

test('ecs - fix log file path', async () => {
    const parser = new ParserEcs()
    const record = {log:{file:{path:'C:\\windows\\file\\path'}}}
    parser.apply(record)
    expect(record).toHaveProperty('log.file.path', 'C:/windows/file/path')
});

test('ecs - referrer - none', async () => {
    const parser = new ParserEcs()
    const record = {http:{request:{referrer:{original:'-'}}}}
    parser.apply(record)
    expect(record).not.toHaveProperty('http.request.referrer.original')
    expect(record).not.toHaveProperty('http.request.referrer.scheme')
    expect(record).not.toHaveProperty('http.request.referrer.domain')
    expect(record).not.toHaveProperty('http.request.referrer.path')
    expect(record).not.toHaveProperty('http.request.referrer.port')
    expect(record).not.toHaveProperty('http.request.referrer.username')
    expect(record).not.toHaveProperty('http.request.referrer.query')
    expect(record).not.toHaveProperty('http.request.referrer.fragment')
    //await expect(record).toMatchSnapshot('9cf7ae9d-2410-464e-9ec6-65d3241dbe8a')
});

test('ecs - referrer - simple', async () => {
    const parser = new ParserEcs()
    const record = {http:{request:{referrer:{original:'http://somewhere.com/with/some/path.ext'}}}}
    parser.apply(record)
    expect(record).toHaveProperty('http.request.referrer.original')
    expect(record).toHaveProperty('http.request.referrer.scheme')
    expect(record).toHaveProperty('http.request.referrer.domain')
    expect(record).toHaveProperty('http.request.referrer.port')
    expect(record).toHaveProperty('http.request.referrer.path')
    expect(record).not.toHaveProperty('http.request.referrer.username')
    expect(record).not.toHaveProperty('http.request.referrer.query')
    expect(record).not.toHaveProperty('http.request.referrer.fragment')
    //await expect(record).toMatchSnapshot('9cf7ae9d-2410-464e-9ec6-65d3241dbe8a')
});

test('ecs - referrer - https', async () => {
    const parser = new ParserEcs()
    const record = {http:{request:{referrer:{original:'https://somewhere.com/with/some/path.ext'}}}}
    parser.apply(record)
    expect(record).toHaveProperty('http.request.referrer.original')
    expect(record).toHaveProperty('http.request.referrer.scheme')
    expect(record).toHaveProperty('http.request.referrer.domain')
    expect(record).toHaveProperty('http.request.referrer.port')
    expect(record).toHaveProperty('http.request.referrer.path')
    expect(record).not.toHaveProperty('http.request.referrer.username')
    expect(record).not.toHaveProperty('http.request.referrer.query')
    expect(record).not.toHaveProperty('http.request.referrer.fragment')
    //await expect(record).toMatchSnapshot('9cf7ae9d-2410-464e-9ec6-65d3241dbe8a')
});

test('ecs - referrer - ws', async () => {
    const parser = new ParserEcs()
    const record = {http:{request:{referrer:{original:'ws://somewhere.com/with/some/path.ext'}}}}
    parser.apply(record)
    expect(record).toHaveProperty('http.request.referrer.original')
    expect(record).toHaveProperty('http.request.referrer.scheme')
    expect(record).toHaveProperty('http.request.referrer.domain')
    expect(record).toHaveProperty('http.request.referrer.port')
    expect(record).toHaveProperty('http.request.referrer.path')
    expect(record).not.toHaveProperty('http.request.referrer.username')
    expect(record).not.toHaveProperty('http.request.referrer.query')
    expect(record).not.toHaveProperty('http.request.referrer.fragment')
    //await expect(record).toMatchSnapshot('9cf7ae9d-2410-464e-9ec6-65d3241dbe8a')
});

test('ecs - referrer - wss', async () => {
    const parser = new ParserEcs()
    const record = {http:{request:{referrer:{original:'ws://somewhere.com/with/some/path.ext'}}}}
    parser.apply(record)
    expect(record).toHaveProperty('http.request.referrer.original')
    expect(record).toHaveProperty('http.request.referrer.scheme')
    expect(record).toHaveProperty('http.request.referrer.domain')
    expect(record).toHaveProperty('http.request.referrer.port')
    expect(record).toHaveProperty('http.request.referrer.path')
    expect(record).not.toHaveProperty('http.request.referrer.username')
    expect(record).not.toHaveProperty('http.request.referrer.query')
    expect(record).not.toHaveProperty('http.request.referrer.fragment')
    //await expect(record).toMatchSnapshot('9cf7ae9d-2410-464e-9ec6-65d3241dbe8a')
});

test('ecs - referrer - wss', async () => {
    const parser = new ParserEcs()
    const record = {http:{request:{referrer:{original:'ftp://somewhere.com/with/some/path.ext'}}}}
    parser.apply(record)
    expect(record).toHaveProperty('http.request.referrer.original')
    expect(record).toHaveProperty('http.request.referrer.scheme')
    expect(record).toHaveProperty('http.request.referrer.domain')
    expect(record).toHaveProperty('http.request.referrer.port')
    expect(record).toHaveProperty('http.request.referrer.path')
    expect(record).not.toHaveProperty('http.request.referrer.username')
    expect(record).not.toHaveProperty('http.request.referrer.query')
    expect(record).not.toHaveProperty('http.request.referrer.fragment')
    //await expect(record).toMatchSnapshot('9cf7ae9d-2410-464e-9ec6-65d3241dbe8a')
});

test('ecs - referrer - wss', async () => {
    const parser = new ParserEcs()
    const record = {http:{request:{referrer:{original:'FiSh://somewhere.com/with/some/path.ext'}}}}
    parser.apply(record)
    expect(record).toHaveProperty('http.request.referrer.original')
    expect(record).toHaveProperty('http.request.referrer.scheme', 'fish')
    expect(record).toHaveProperty('http.request.referrer.domain', 'somewhere.com')
    expect(record).toHaveProperty('http.request.referrer.path', '/with/some/path.ext')
    expect(record).not.toHaveProperty('http.request.referrer.port')
    expect(record).not.toHaveProperty('http.request.referrer.username')
    expect(record).not.toHaveProperty('http.request.referrer.query')
    expect(record).not.toHaveProperty('http.request.referrer.fragment')
    //await expect(record).toMatchSnapshot('9cf7ae9d-2410-464e-9ec6-65d3241dbe8a')
});

test('ecs - referrer - no path', async () => {
    const parser = new ParserEcs()
    const record: any = {http:{request:{referrer:{original:'http://somewhere.com'}}}}
    const matches = parser.matches(record)
    parser.apply(record)
    expect(matches).toEqual(true)
    expect(record).toHaveProperty('http.request.referrer.original')
    expect(record).toHaveProperty('http.request.referrer.scheme')
    expect(record).toHaveProperty('http.request.referrer.domain')
    expect(record).toHaveProperty('http.request.referrer.port')
    expect(record).toHaveProperty('http.request.referrer.path')
    expect(record.http.request.referrer.path).toMatch(/^\//)
    expect(record).not.toHaveProperty('http.request.referrer.username')
    expect(record).not.toHaveProperty('http.request.referrer.query')
    expect(record).not.toHaveProperty('http.request.referrer.fragment')
    await expect(record).toMatchSnapshot('fe3a90a4-1319-44e0-af0f-312686d81ae7')
});

test('ecs - referrer - path with just /', async () => {
    const parser = new ParserEcs()
    const record:any = {http:{request:{referrer:{original:'http://somewhere.com/'}}}}
    parser.apply(record)
    expect(record).toHaveProperty('http.request.referrer.original')
    expect(record).toHaveProperty('http.request.referrer.scheme')
    expect(record).toHaveProperty('http.request.referrer.domain')
    expect(record).toHaveProperty('http.request.referrer.port')
    expect(record.http.request.referrer.port).toEqual('80')
    expect(record).toHaveProperty('http.request.referrer.path')
    expect(record.http.request.referrer.path).toMatch(/^\//)
    expect(record).not.toHaveProperty('http.request.referrer.username')
    expect(record).not.toHaveProperty('http.request.referrer.query')
    expect(record).not.toHaveProperty('http.request.referrer.fragment')
    await expect(record).toMatchSnapshot('84023bc4-77fe-439f-9123-91c20362c04a')
});

test('ecs - referrer - full', async () => {
    const parser = new ParserEcs()
    const record: any = {http:{request:{referrer:{original:'http://username:password@somewhere.com:8080/with/some/path.ext?this=that#fragment'}}}}
    parser.apply(record)
    expect(record.http.request.referrer.port).toEqual('8080')
    await expect(record).toMatchSnapshot('00abcd8b-a685-44ab-a5ef-b43ce489d6cd')
});

test('ecs - http.request.line - 01', async () => {
    const parser = new ParserEcs()
    const record = {http:{request:{line:"GET /?XDEBUG_SESSION_START=phpstorm HTTP/1.1"}}, url:{scheme: 'https', port: "443", domain: 'localhost'}}
    parser.apply(record)
    expect(record).toHaveProperty('http.request.method', 'GET')
    expect(record).toHaveProperty('http.version', '1.1')
    expect(record).toHaveProperty('url.original', '/?XDEBUG_SESSION_START=phpstorm')
    expect(record).toHaveProperty('url.path', '/')
    expect(record).toHaveProperty('url.query', 'XDEBUG_SESSION_START=phpstorm')
});

test('ecs - http.request.line - 02', async () => {
    const parser = new ParserEcs()
    const record = {http:{request:{line:"GET /?XDEBUG_SESSION_START=phpstorm#fragment HTTP/1.1"}}, url:{scheme: 'https', port: "443", domain: 'localhost'}}
    parser.apply(record)
    expect(record).toHaveProperty('http.request.method', 'GET')
    expect(record).toHaveProperty('http.version', '1.1')
    expect(record).toHaveProperty('url.original', '/?XDEBUG_SESSION_START=phpstorm#fragment')
    expect(record).toHaveProperty('url.path', '/')
    expect(record).toHaveProperty('url.query', 'XDEBUG_SESSION_START=phpstorm')

});

test('ecs - http.request.line - 03', async () => {
    const parser = new ParserEcs()
    const record = {http:{request:{line:"POST /ext/jcrs/rest_v2/import?update=false HTTP/1.1"}}, url:{scheme: 'https', port: "443", domain: 'localhost'}}
    parser.apply(record)
    expect(record).toHaveProperty('http.request.method', 'POST')
    expect(record).toHaveProperty('http.version', '1.1')
    expect(record).toHaveProperty('url.original', '/ext/jcrs/rest_v2/import?update=false')
    expect(record).toHaveProperty('url.path', '/ext/jcrs/rest_v2/import')
    expect(record).toHaveProperty('url.query', 'update=false')
});

test('ecs - http.request.line - 04', async () => {
    const parser = new ParserEcs()
    const record = {http:{request:{line:"GET /ext/farm/farm265.do;jsessionid=JA6pko6NolLG_MPnJNoBbx75a-aAV9x99zjo0ECs1tH-VVS0waj4!1029141353 HTTP/1.1"}}, url:{scheme: 'https', port: "443", domain: 'localhost'}}
    parser.apply(record)
    expect(record).toHaveProperty('http.request.method', 'GET')
    expect(record).toHaveProperty('http.version', '1.1')
    expect(record).toHaveProperty('url.original', '/ext/farm/farm265.do;jsessionid=JA6pko6NolLG_MPnJNoBbx75a-aAV9x99zjo0ECs1tH-VVS0waj4!1029141353')
    expect(record).toHaveProperty('url.path', '/ext/farm/farm265.do')
    expect(record).toHaveProperty('url.path_param', 'jsessionid=JA6pko6NolLG_MPnJNoBbx75a-aAV9x99zjo0ECs1tH-VVS0waj4!1029141353')
    expect(record).toHaveProperty('url.file.directory', '/ext/farm')
    expect(record).not.toHaveProperty('url.query')
});

test('ecs - http.request.line - 05', async () => {
    const parser = new ParserEcs()
    const record = {http:{request:{line:"GET /pub/oauth2/v1/oauth/token?disableDeveloperFilter=true&grant_type=client_credentials&scope=ACTIVEMQ.* HTTP/1.1"}}, url:{scheme: 'https', port: "443", domain: 'localhost'}}
    parser.apply(record)
    expect(record).toHaveProperty('http.request.method', 'GET')
    expect(record).toHaveProperty('http.version', '1.1')
    expect(record).toHaveProperty('url.original', '/pub/oauth2/v1/oauth/token?disableDeveloperFilter=true&grant_type=client_credentials&scope=ACTIVEMQ.*')
    expect(record).toHaveProperty('url.path', '/pub/oauth2/v1/oauth/token')
    expect(record).toHaveProperty('url.file.directory', '/pub/oauth2/v1/oauth')
    expect(record).toHaveProperty('url.query', 'disableDeveloperFilter=true&grant_type=client_credentials&scope=ACTIVEMQ.*')
});

test('ecs - http.request.line - 06', async () => {
    const parser = new ParserEcs()
    const record = {http:{request:{line:"GET /pub/oauth2/v1/oauth/token?disableDeveloperFilter=true&grant_type=client_credentials&scope=DSP.* HTTP/1.1"}}, url:{scheme: 'https', port: "443", domain: 'localhost'}}
    parser.apply(record)
    expect(record).toHaveProperty('http.request.method', 'GET')
    expect(record).toHaveProperty('http.version', '1.1')
    expect(record).toHaveProperty('url.original', '/pub/oauth2/v1/oauth/token?disableDeveloperFilter=true&grant_type=client_credentials&scope=DSP.*')
    expect(record).toHaveProperty('url.path', '/pub/oauth2/v1/oauth/token')
    expect(record).toHaveProperty('url.query', 'disableDeveloperFilter=true&grant_type=client_credentials&scope=DSP.*')
});
test('ecs - http.request.line - 07', async () => {
    const parser = new ParserEcs()
    const record = {http:{request:{line:"GET /ext/farm/farm265.do; HTTP/1.1"}}, url:{scheme: 'https', port: "443", domain: 'localhost'}}
    parser.apply(record)
    expect(record).toHaveProperty('http.request.method', 'GET')
    expect(record).toHaveProperty('http.version', '1.1')
    expect(record).toHaveProperty('url.original', '/ext/farm/farm265.do;')
    expect(record).toHaveProperty('url.path', '/ext/farm/farm265.do')
    expect(record).not.toHaveProperty('url.path_param')
    expect(record).not.toHaveProperty('url.query')
});
test('ecs - http.request.line - 08', async () => {
    const parser = new ParserEcs()
    const record = {http:{request:{line:'GET /WebID/IISWebAgentIF.dll?postdata=\\"><script>foo</script> HTTP/1.1'}}, url:{scheme: 'https', port: "443", domain: 'localhost'}}
    parser.apply(record)
    expect(record).toHaveProperty('http.request.method', 'GET')
    expect(record).toHaveProperty('http.version', '1.1')
    expect(record).toHaveProperty('url.original', '/WebID/IISWebAgentIF.dll?postdata=\\"><script>foo</script>')
    expect(record).toHaveProperty('url.path', '/WebID/IISWebAgentIF.dll')
    expect(record).not.toHaveProperty('url.path_param')
    expect(record).toHaveProperty('url.query', 'postdata=\\%22%3E%3Cscript%3Efoo%3C/script%3E')
});