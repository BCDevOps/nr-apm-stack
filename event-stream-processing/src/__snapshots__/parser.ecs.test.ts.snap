// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ecs - referrer - full: 00abcd8b-a685-44ab-a5ef-b43ce489d6cd 1`] = `
Object {
  "http": Object {
    "request": Object {
      "referrer": Object {
        "domain": "somewhere.com",
        "file": Object {
          "directory": "/with/some",
          "ext": "ext",
          "name": "path.ext",
        },
        "fragment": "fragment",
        "full": "http://somewhere.com:8080/with/some/path.ext?this=that#fragment",
        "original": "http://username:password@somewhere.com:8080/with/some/path.ext?this=that#fragment",
        "password": "-",
        "path": "/with/some/path.ext",
        "port": "8080",
        "query": "this=that",
        "query_k": Array [
          "this",
        ],
        "query_kv": Array [
          "this=that",
        ],
        "query_v": Array [
          "that",
        ],
        "scheme": "http",
        "uri": "//somewhere.com/with/some/path.ext",
        "username": "username",
      },
    },
  },
}
`;

exports[`ecs - referrer - no path: fe3a90a4-1319-44e0-af0f-312686d81ae7 1`] = `
Object {
  "http": Object {
    "request": Object {
      "referrer": Object {
        "domain": "somewhere.com",
        "file": Object {
          "directory": "/",
        },
        "full": "http://somewhere.com/",
        "original": "http://somewhere.com",
        "path": "/",
        "port": "80",
        "scheme": "http",
        "uri": "//somewhere.com/",
      },
    },
  },
}
`;

exports[`ecs - referrer - path with just /: 84023bc4-77fe-439f-9123-91c20362c04a 1`] = `
Object {
  "http": Object {
    "request": Object {
      "referrer": Object {
        "domain": "somewhere.com",
        "file": Object {
          "directory": "/",
        },
        "full": "http://somewhere.com/",
        "original": "http://somewhere.com/",
        "path": "/",
        "port": "80",
        "scheme": "http",
        "uri": "//somewhere.com/",
      },
    },
  },
}
`;
